### DISCUSSION TITLE
Localization notes and tips

### DISCUSSION THREAD

**Discussion by BlakeNeko** [sandboxie-plus/Sandboxie#1123] [Documents and notes]
When I'm using Sandboxie-plus, I noticed some zh-CN translation is missing or can be improved.
I tried walk through the source code and found the translation files but don't know how to get start.
Since I don't have any experience how to add/modify translations while using github or other tools, it's Is there any methods I can help to fix them?
(Not a native English speaker, sorry if I made some grammar mistakes)

**Comment by isaak654** [comment]
<div align="center">
<h2>How to locate translations in the Sandboxie repository</h2>
</div>

<ol>
<li>
<a href="https://github.com/sandboxie-plus/Sandboxie/blob/master/SandboxiePlus/SandMan/">Here</a> you can find the location of the Sandboxie Plus translation files in the repository (<i>sandman_*.ts</i>)
</li>
<li>
<a href="https://github.com/sandboxie-plus/Sandboxie/blob/master/Installer/Languages.iss">Here</a> you can find the location of the Sandboxie Plus installer translation file (recommended if you have already translated Plus interface)
</li>
<li>
<a href="https://github.com/sandboxie-plus/Sandboxie/blob/6e830b9c0a00588f72120d9db2ef559bc76e210b/Installer/Sandboxie-Plus.iss#L222-L242">Here</a> you can find the location of the Sandboxie Plus installer script with the language value to add (recommended if you have already translated Plus interface)
</li>
<li>
<a href="https://github.com/sandboxie-plus/Sandboxie/tree/master/SandboxiePlus/SandMan/Troubleshooting">Here</a> you can find the location of the Sandboxie Plus translation files for the troubleshooting wizard (<i>lang_*.json</i>)
</li>
<li>
<a href="https://github.com/sandboxie-plus/Sandboxie/blob/55cf24f903e4833bbaf3fc33d6ff03185819b0ef/Sandboxie/install/SbieSettings.ini#L81-L244">Here</a> you can find the location of the Sandboxie Plus localized strings for <i>SbieSettings.ini</i>
</li>
<li>
<a href="https://github.com/sandboxie-plus/Sandboxie/blob/master/Sandboxie/msgs/">Here</a> you can find the location of the Sandboxie Classic translation files (<i>Sbie-English-1033.txt</i>, <i>Text-*-*.txt</i>)
</li>
<li>
<a href="https://github.com/sandboxie-plus/Sandboxie/blob/master/Sandboxie/install/SandboxieVS.nsi">Here</a> you can find the location of the Sandboxie Classic installer script (in case you have to provide a missing language)
</li>
</ol>

<div align="center">
Please consider to create new pull requests periodically if you have new language updates to merge.
For this task, I would recommend to use <a href="https://desktop.github.com/">GitHub desktop</a> because it helps to manage a lot of strings. Github Desktop for Linux-based systems is not officially supported, but there is an available <a href="https://github.com/shiftkey/desktop">fork</a>.
</div>

<a name="lupdate_win"></a>
<div align="center">
<h2>Qt Linguist to update Plus translations</h2>
</div>

The official translation tool for Sandboxie Plus is `Qt Linguist`, which usually requires the installation of the [Qt online installer](https://www.qt.io/download-qt-installer) (online course: [here](https://courses.qt.io/how-to-install-qt/index.html?hsLang=en#/)). At the end of the process, you will be able to run the following command lines that I found useful to sync and translate the missing strings:

`C:\Qt\6.1.1\msvc2019_64\bin\lupdate.exe SandMan.pri` at `\SandboxiePlus\SandMan` (shared by [yfdyh000](https://github.com/sandboxie-plus/Sandboxie/pull/905#issuecomment-882049888))
or
 `C:\Qt\6.1.1\msvc2019_64\bin\lupdate-pro.exe SandboxiePlus.pro` at `\SandboxiePlus`.
 
>  [!NOTE]
> If you need to remove obsolete entries that are no longer used, just add the `-no-obsolete` argument to the lupdate command.
> For a full list of lupdate options, see https://doc.qt.io/qt-6/linguist-lupdate.html

You would need to make a local clone of the repository first (for example with the Github Desktop tool mentioned before), then running one of the commands above from the path where SandMan.pri or SandboxiePlus.pro is located (it depends on the command you have chosen). In this way, you'll get the new changes in GitHub Desktop (or in any Git client you're using). At this point, you can use a translation tool like Qt Linguist to update the translation.

At the beginning of the output, you may receive a number of `Cannot run compiler 'cl'` , but all current .ts files will be still updated.

<div align="center">
<h2>Using Qt Creator to synchronize Sandboxie Plus translations</h2>
</div>

In order to set up the sync on the `Qt Creator` tool, select the menu Tools -> External -> Configure -> Environment -> External Tools -> Update Translations (lupdate). For more information, see also [Using External Tools](https://doc.qt.io/qtcreator/creator-editor-external.html).

In my configuration, the executable is set to `C:\Qt\5.15.2\msvc2019_64\bin\lupdate-pro.exe`, the Arguments line needs to contain just `SandboxiePlus.pro` and the working directory needs the full path which contains SandboxiePlus.pro and all other repository folders.
After doing that, if everything is correct, just run `Update Translations (lupdate)` from Tools menu -> External -> Linguist

:warning: **Note:** if the executable is found in your system PATH variable, do not specify the path to it.

<div align="center">
<h2>Set the target language and country for Plus translations</h2>
</div>

<p align="center">
Once your `sandman_xx.ts` file is open in `Qt Linguist`, you can customize the locale information from the Edit menu -> Translation File Settings -> set one of the values in the Target language category -> save the .ts file from the File menu. 
For more information, see also <a href="https://doc.qt.io/qt-6/linguist-translators.html#changing-the-target-locale">Changing the Target Locale</a>.
</p>

<div align="center">
<h2>Recommendations for all translators in this repository</h2>
</div>

<p align="center"> :warning: You shouldn't ever edit the text inside the source tags of the .ts file(s) before applying the same change on the respective source code file(s). For example, in commit <a title="Commit to remove useless space" href="https://github.com/sandboxie-plus/Sandboxie/commit/3b864411275ac0fa6e60ea1544feb176b470c049">3b86441</a> I edited the source file and all the .ts files together. In your case, you could just update your specific .ts file together with the source file(s) in question and nothing else.</p>

<p align="center"> :warning: As a rule of thumb, you shouldn't ever translate the .ts file(s) by hand, in order to avoid formatting issues (the use of an official translation tool like <code>Qt Linguist</code> is highly recommended).</p>

<p align="center"> :warning: It's not useful to reintroduce the translated entries tagged as vanished, because they're outdated. It's recommended to keep them as they are for a limited time as reference, in order to avoid future clutter. A similar recommendation could be applied for the oldsource tags.</p>

<p align="center">:bookmark_tabs: There is a new scheduled job to synchronise the language entries that are updated or added in new Sandboxie releases. See <a href="https://github.com/sandboxie-plus/Sandboxie/issues/1595#issuecomment-1168581852">A new workflow to improve delivery of language updates</a> for a quick explanation of how it works.</p>

<p align="center"> :bookmark_tabs: Once the source files for Sandboxie Plus are updated, the <a href="https://github.com/sandboxie-plus/Sandboxie/blob/master/.github/workflows/lupdate.yml">lupdate workflow</a> takes care of updating all .ts files regularly through a schedule, for example: <a title="Automatic commit to update Sandboxie Plus language entries" href="https://github.com/sandboxie-plus/Sandboxie/commit/d11000f23f6e0968060356c2ab9ab4bbc7eabc67">d11000f</a>.</p>

<p align="center"> :bookmark_tabs: Sandboxie Classic translation files are also used in specific Plus options, so it's advisable to look at the reports <a href="https://github.com/sandboxie-plus/Sandboxie/tree/master/Sandboxie/msgs/report">here</a> to track the new entries. </p>

</div>

<div align="center">
<h2>Further translation tips</h2>
</div>

1) Steps to install/set Qt:
https://github.com/sandboxie-plus/Sandboxie/issues/885#issuecomment-858502218
2) How to use Qt Linguist to prevent formatting issues:
https://github.com/sandboxie-plus/Sandboxie/pull/2133#pullrequestreview-1072117923
3) How to test the translation locally in Sandboxie Plus:
https://github.com/sandboxie-plus/Sandboxie/issues/593#issuecomment-880562063
4) AutoIt3 script that gets the Sandboxie Classic strings for the Plus Installer custom messages in Languages.iss:
https://github.com/sandboxie-plus/Sandboxie/pull/1455#issuecomment-999992715

<div align="center">
<h2>Building Qt 5 sources (for developers)</h2>
</div>

<div align="center">

https://wiki.qt.io/Building_Qt_5_from_Git
https://doc.qt.io/qt-5/linux-building.html
https://www.linuxfromscratch.org/blfs/view/svn/x/qt5.html
https://stackoverflow.com/q/14571701
https://stackoverflow.com/q/14170590

</div>

**Reply by sebadamus** [reply]
Newbie question... just to be sure... I do have to download the project locally to do the sync stuff, right?
I mean... C:\Qt\6.1.1\msvc2019_64\bin\lupdate.exe SandMan.pri?

Thanks,

PS: just got the QT complete running with wine, and the lupdate.exe found! (I had previously downloaded just QT-Linguist and hasnt the lupdate.exe

**Reply by isaak654** [reply]
You need to make a local clone of the repository first, then running the command from the path where SandMan.pri is located. In this way, you'll get the new changes in GitHub Desktop (or any Git client you're using). At this point, you can use Qt Linguist to update the translation.

**Reply by sebadamus** [reply]
Thanks, just could lupdate my sandman_es.ts there are about 20 extra stuff to translate, will do soon!

**Comment by Zerorigin** [comment]
This is my use case.  I hope it will be helpful for others.
```
# fork & clone the repo
pushd R:/
git clone --depth 1 git@github.com:[your_username]/Sandboxie.git

# Download [Qt Linguist Standalone](https://github.com/lelegard/qtlinguist-installers/releases), it's sufficien for the job.
# sync & lupdate single ts file example
pushd R:\Sandboxie\SandboxiePlus\SandMan
D:\Devs\QtLinguist\lupdate.exe . -ts sandman_zh_CN.ts
D:\Devs\QtLinguist\linguist.exe .\sandman_zh_CN.ts
　
```

**Reply by Zerorigin** [reply]
It's easy to create a workflow, but some translators want to keep special formatted of source lines at .ts file, which is a trouble.

**Reply by isaak654** [reply]
See https://stackoverflow.com/a/34903961



> It's part of the XML language definition. Since `'` is a special character in xml, it has to be escaped. Otherwise, a XML parser will not be able to read the file. In other words `&apos;` is nothing but an escape sequence for `'` and any XML reader will replace `&apos;` with `'` while reading the xml file.
> 
> The escape sequences are:
> 
> ```
> "   &quot;
> '   &apos;
> <   &lt;
> >   &gt;
> &   &amp;
> ```



Any .ts file starts with `<?xml version="1.0" encoding="utf-8"?>`

**Comment by RimacC2-EV** [comment]
Hello everyone there, When I'm translating Sandboxie, I found some `<translation type="vanished">***<translation>` and `<translation type="unfinished">***<translation>` which seems to make the .ts file fat, can(some of) them being deleted?

**Reply by isaak654** [reply]
You can add the `-no-obsolete` argument to the lupdate command(s) mentioned above.
For a full list of lupdate options, see https://doc.qt.io/qt-6/linguist-lupdate.html
